src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs

```diff
@@ -1098,6 +1099,33 @@ private BoundExpression MakeNullableHasValue(SyntaxNode syntax, BoundExpression
                 sideEffects: ImmutableArray.Create<BoundExpression>(tempAssignmentX, tempAssignmentY),
                 value: binaryExpression,
                 type: boolType);
+
+            static bool canNotBeEqualToDefaultValue(
+                [NotNullWhen(returnValue: true)] ConstantValue? constantValue)
+            {
+                // This is an explicit list so new constant values are not accidentally supported without consideration.
+                // Decimal is not in the list because it is possible to have a non-default decimal constant
+                // which is equal to the default decimal (0.0m == default(decimal)).
+                return constantValue is
+                {
+                    IsDefaultValue: false,
+                    Discriminator: ConstantValueTypeDiscriminator.Boolean
+                        or ConstantValueTypeDiscriminator.Byte
+                        or ConstantValueTypeDiscriminator.Char
+                        or ConstantValueTypeDiscriminator.Double
+                        or ConstantValueTypeDiscriminator.Int16
+                        or ConstantValueTypeDiscriminator.Int32
+                        or ConstantValueTypeDiscriminator.Int64
+                        or ConstantValueTypeDiscriminator.NInt
+                        or ConstantValueTypeDiscriminator.NUInt
+                        or ConstantValueTypeDiscriminator.SByte
+                        or ConstantValueTypeDiscriminator.Single
+                        or ConstantValueTypeDiscriminator.String
```

Comment:
>ConstantValueTypeDiscriminator.String [](http://example.com/codeflow?start=27&length=37)

Consider removing `String` if this local function should only be called with value types.